### ISSUE TITLE
late registration

### ISSUE THREAD

**Issue by ctm** [ctm/mb2-doc#183]
Allow tournaments to have an optional late registration period during which new players can join a running tournament.

This issue is for the minimally viable late registration which doesn't split tables to create new empty seats, nor does it have a queue of people waiting to join. However, there are still a number of fairly big changes that need to be made to support late registration, some of which I overlooked when I started working on this issue.

This issue has been tricky enough that I've rewritten this description and will continue to add (and eventually check off) check-boxes as needed until I get it done.

The following things have either been done or need to be done. As I run into more things that need to be done, this list will expand.
- [x] In create, use `BettingRoundOrdinal` notation
- [x] Figure out how synchronized breaks affect the new `BettingRoundOrdinal` code[^1]
- [x] ~~Use same notation as we currently use for end of rebuy~~
- [x] Rename `UpcomingParent` and make it a type alias (i.e., get rid of the bad practices before duplicating upcoming.rs)
- [x] Refactor the `Upcoming` component to be `Events` on the outside and `Upcoming` on the inside
- [x] Create `Running` component (start with upcoming.rs) and add it to `Events`
- [x] Update Tournament to detect end of late start (i.e., check when we update level)
  - [x] Make sure the last late registration level is shoehorned into `TournamentState` (`State::new` takes a `&Structure` parameter, but the last late registration level is a parameter and not part of `Structure`)
- [x] Don't nuke lobby listing until after end of late start (Tournament sends end message)
- [x] Create various errors
- [x] Create separate issue for splitting tables to accommodate late entry
- [x] Allow the join checkbox after the tournament has started for players not already in
- [x] Add player when seat is available
  - [x] review / reuse rebuy code to make sure FM is collected and all database tables are updated
- [x] Address n-player and prize pool increase
  - [x] Decide if we want to continue to allow rebuys after a place has been paid
  - [x] Increase the prize pool when someone late joins
  - [x] Figure out if the number of places to pay changes when someone joins
- [x] Keep the tournament in nickmapper as long as late registrations are still allowed, so new logins see it[^2]
- [x] Keep the `entrant_ids` field current in the running events in the lobby
  - [x] Have tournament send message when someone joins
  - [x] Have lobby process that join message
    - [x] Make nickmapper know how to map table ids to event ids
      - [x] Add an `EventId` to the `tables` HashMap
      - [x] Create and send a `TablesAdded` message once the TableIds are known
      - [x] insert when events are created
      - [x] insert when tables are created
      - [x] delete when tables are destroyed
      - [x] delete when tables are destroyed
    - [x] process Public::PlayersAdded
      - [x] tournament
      - [x] ring game (maybe&hellip; ring game support is incomplete)
  - [x] Have something (tournament or table) send message when someone busts, quits or wins (there's a `Won` message that's already sent when FunMoney is won, but nothing when someone finishes without winning)
  - [x] Have lobby process that bust/quit/win message
- [x] When a player logs in, send one or more messages to populate the running info that already logged in people have
- [x] Refactor RunningEvents into RunningEvent
  - [x] update the message name and contents in pen_message.rs
  - [x] process the RunningEvent message in lobby.rs
  - [x] in nick_mapper::announce_running_events, ask the running tournament to contact the user
  - [x] have the running tournament contact the user
  - [x] ~~when a tournament is started, send a RunningEvent message to everyone (eventually this will be gated by some sort of predicate, but we don't need to worry about that scale yet)~~ (no, restore and rehabilitate the code that moves events from upcoming to running)
- [x] Restore and rehabilitate the code that moves events from upcoming to running
  - [x] Start with LateRegistrationUnavailable(entries: 0, entries_left: 1, currently_playing: 0)
  - [x] Update that status when PlayersAdded gets sent (if it is our player id, then increment entries, decrement entries_left and set currently playing to true)
  - [x] Then, on start, change it to LateRegistrationAvailable if there are slots ...
- [x] Update running.rs and running/event_row.rs to use the registration_state field
  - [x] Checkbox should only be enabled when appropriate
  - [x] If disabled, tooltip should say why
- [x] Get rid of Running entry when tournament ends
  - [x] Send the status change at end
  - [x] Process the status change
- [x] Allow join checkbox if player not already in (and initially, if there's a seat)
  - [x] Get rid of upcoming entry once a player joins
  - [x] Send RegistrationAvailable at startup if not all tables are full (if late registration is available)
    - [x] Implement
    - [x] Test
  - [x] Handle when a player busts or quits
    - [x] Send RegistrationAvailable on bust (if late registration is still available) unless tables are combined
      - [x] Implement
      - [x] Test
    - [x] Send RegistrationUnavailable if tables are combined and it results in no empty seats
      - [x] Implement
      - [x] Test
  - [x] Send RegistrationUnavailable when someone joins if that fills the tournament (and late registration is still available)
      - [x] Implement
      - [x] Test
- [x] Provide per-player info about joining that takes re-entries into account (not allowing re-entry right now)
- [x] Provide a better explanation of the Late Registration field in the creation form
- [x] Have the lobby summary drop-down provide late registration info
- [x] Send email to the BARGE list
- [x] Make sure that the new player starts in the proper state of either folded or sitting out
  - [x] If sitting out, make sure the player sits out exactly all the hands that the player should sit out (not fewer or more)
- [x] Double check that the last late registration level works properly both with and without breaks on the 55
- [x] Update the [Players' Manual](https://ctm.github.io/docs/players_manual/)

[^1]: The new code is written assuming that Breaks aren't pulled out of Rounds, but IIRC, that's how we implement synchronized Breaks, we take the Level::Break rounds out and then update other BettingRoundIndexes, like the last rebuy round. As such, we'll also need to do the same adjustment for last late registration round. This isn't hard, but it's _ugly_.
[^2]: This is already available in the `tournaments` field. Duh.

**Comment by ctm** [comment]
PS no longer supports late registration, so people are more likely to understand that they need to sign up in time. I think that just getting #86 done should be sufficient.

We have enough feature requests that I'm closing this at least for now.

**Comment by ctm** [comment]
PS has it again, so we need it.

**Comment by ctm** [comment]
I'll consider #238 a @jdkullmann thumbs-up

**Comment by ctm** [comment]
Oops. I had tagged that as won't fix, but forgot to remove the tag, so I did that now.

**Comment by ctm** [comment]
Here's a bunch of useful information from C. Meck:

> Late Registration:
> 
> It's pretty typical in live play for tables to start short-handed (for example, you might start a 9-handed tournament with 6 or 7 players at each table) and be filled in with late registrations until either the maximum capacity allotted to the tournament is reached (at this point, either alternates are taken or players are turned away) or late registration ends.  In the latter case, tables can then be broken in order to have full tables.
> 
> Here's how it works on Stars.
> 
> Suppose we are playing HORSE, with 8 player tables.  Suppose at the start of the tournament 6 players are registered.  Two minutes in Nancy late-regs, and is place at one of the empty seats.  If eight minutes in Sabyl late-regs, since that's the final seat at the table, Stars will instead place that player at a new table and pull players from the existing tables.  In this case, you'd end up with 2 tables of 4 players each.  If late reg then ended with those 8 players, then they are re-combined to a full table.  It's kind of awkward and not really loved by me, so I don't feel that it should necessarily be imitated.
> 
> In a bigger situation, suppose we started that HORSE tournament with 14 players.  They are at 2 full tables of 7.  The first late registrant is randomly placed to one of the empty seats, giving us a full table of 8 and a not-quite full table of 7.  The second late registrant would take the last available seat, so Stars creates a third table at this point with the new player and players brought over from the existing tables.  You'd have tables of 6, 5, & 5 at this point.  Other late registrants are assigned to the shortest table (randomly if multiple tables are equally short).
> 
> I'd rather not create the new table until absolutely necessary (so the 9th player in the HORSE tournament to go from 1 to 2 tables, the 17th player to go from 2 to 3 tables, etc.).
> 
> [snip]
> 
> As far as your scenario--suppose we start a 9 handed tournament with exactly 18 players and late registration available.  Several options.
> 1. Start with 2 full tables, create a third table when someone registers.
> 2. Start with 3 tables of 6, seat late regs at the empty seats, create a 4th table if there are enough late regs to warrant it.
> 
> A bigger situation, suppose you have a 9 handed tournament with exactly 100 players and late registration available.
> 1. Start with tables as full as possible, so since 9*11=99, you need 12 tables.  Have 4 tables of 9 and 8 tables of 8, seat the first 8 late regs at those empty seats, and create a new table or take alternates or close registration once you have 9*12=108 players.
> 2. If you anticipate lots of late regs, start 6 or 7 handed.  If 6 handed, then 100/6=16.6666, so start with 16 tables with 12 tables of 6 and 4 tables of 7.  If 7 handed, then 100/7=14.28, start with 14 tables with 12 tables of 7 and 2 tables of 8.  Fill in tables until full, then create new tables/alternates/close reg.  In an online situation, you'd probably just create new tables because you don't have the physical constraints of available tables/dealers that exists live.
>

**Comment by ctm** [comment]
Dan asked for this just now, so at minimum, I will create a set of check-boxes that show what needs to be done. I think that due to some of the other functionality that I've added over the last couple of years, this should be fairly easy to implement. After all, we already have rebuys.

The only real new thing is splitting tables and we could start out with late registration that never splits tables but simply doesn't allow someone to register late unless there's an empty seat somewhere. But even splitting tables shouldn't be hard.

I still have a lot of business stuff to work on, but this is something I can work on when I'm tired (although in theory, I've made some lifestyle changes so I'll be tired less).

**Comment by ctm** [comment]
Oh, and although I'd do this just because Dan asked, this also might help us get a few more players between now and when I turn ring games back on.

**Comment by ctm** [comment]
I added `high priority` almost two weeks ago and never touched this. I will do some of the checkboxes today and either get it going or at least get it to a stable point (and probably merge it into master).

**Comment by ctm** [comment]
Other issues took precedence over this one during the last few days, notably the narcoleptic dinosaur (#1060) and the seat position after player motion bug (#1058). I'm not going to be able to work on this one for a little while.

**Comment by ctm** [comment]
I've been slowly working on this here and there, and I realized that if we're using synchronized breaks, then we should let them specify the last level, rather than the last round. Doing it this way makes for a better user experience, but does require a bit of hoop-jumping on our part. It doesn't help that internally rounds are zero based and levels are one based.

**Comment by ctm** [comment]
Best solution is to use levels for both sync'd breaks and non-syncs in the UI, but to have those converted to rounds before they go into the database. That's still slightly tricky because it's not clear how to represent something in the UI that comes from the database but is not choosable in the UI (i.e., choosing to end the late registration at the end of a break), but that issue was already present with what I had done so far.

**Comment by ctm** [comment]
So, the reason I had gone with rounds, rather than levels initially is that levels technically have names, although off the top of my head, currently only our TOC-style tournaments use anything other than "Level 1", "Level 2", etc. TOC-style has "Level 1a", "Level 1b", "Level 2a", etc. Currently, only the server knows these names in advance, which means without some refactoring we can't say "Level 4" for regular tournaments and "Level 2b" for TOC-style tournaments to represent the fourth level, but we can say "fourth level" or even "Fourth Level" and convey the same information, albeit in a sub-optimal form.

**Comment by ctm** [comment]
I think reusing upcoming to represent in-progress tournaments is not a hack worth doing after all. Although it will take a little work and I'll have to refactor some of what I've already done, having a separate data structure for events that are currently running is better than trying to shoehorn both upcoming and running tournaments into a single data structure.

FWIW, although this is more work to do before I can finish late-registration, it's work that'll need to be done eventually anyway, since we'll want observers to be able to look at tournaments in progress and see things like the number of players. It's been almost three years since I created this issue, so I can't be sure, but my guess is I didn't initially want to do it this way due to it requiring more front-end work. As a compromise, I can make it so that nickmapper keeps track of more stuff than it passes to the lobby and the lobby receives more information than it displays.

**Comment by ctm** [comment]
Although I never wrote a comment mentioning it, I did do the refactor I mentioned in my last comment. So, at least in theory, although there are many more tasks for me to complete, none of them are particularly involved. Other than fixing bugs, getting (and keeping) the WSOPS series running and various business tasks, this is my highest priority now.

**Comment by ctm** [comment]
FWIW, I've been working on this a little bit almost every day since I got back from BARGE. I'm going to try to get it at least limping before leaving for New York on Tuesday the 8th.

I haven't been updating the checkboxes, because I've done some refactoring and would rather work on the code than spend time reorganizing the checkboxes.

**Comment by ctm** [comment]
FWIW, I deployed my `late-183` branch and was getting ready to post to the BARGE list when I realized that the Late Registration field of the creation dialog doesn't explicitly state (not even in a tooltip) that the level being chosen is the last level to allow late registration (as opposed to the first level where late registration isn't allowed). Furthermore&mdash;and worse&mdash;the event drop-down summary doesn't even mention if late registration is allowed, So&hellip;

I've added the above as checkboxes and will not post to the BARGE list until they're done, but I really have some non-programming that I have to do, so it's an annoying situation to be in, albeit much better than the never-ending checkboxes.
